### ISSUE TITLE
[BUG] SSL error even after disabling verification of SSL security

### ISSUE THREAD

**Issue by Poorchop** [FreshRSS/FreshRSS#3752]
**Describe the bug**
Some of my feeds use self-signed certificates. When I set "Verify SSL security" to "No" for these feeds individually, I am still not able to access the feeds via FreshRSS. I get the following error:

```
cURL error 59: failed setting cipher list: DEFAULT@SECLEVEL=1 [https://www.example.com]
```

**To Reproduce**
Some of the feeds are on my own server using a self-certificate that I generated myself using the following:

```
openssl req -x509 -nodes -days 365 -newkey rsa:2048
```

The feed is password-protected so I set the correct username and password in FreshRSS, change "Verify SSL security" to "No," and I try to refresh the feed but it fails with the above error.

**Expected behavior**
FreshRSS should disregard any SSL errors and fetch the feed.

**Environment information (please complete the following information):**
 - OS: OpenBSD 6.9
 - Browser: Firefox 90.0.2
 - FreshRSS version: 1.18.1
 - Database version: PostgreSQL 13.2
 - PHP version: PHP 8.0.9
 - Installation type: Self-hosted server

**Additional context**
 - PHP-FPM runs in a chrooted environment by default in OpenBSD. This led to some php-curl errors for me until I figured out that I need to copy `/etc/resolv.conf` and `/etc/ssl/` to the chroot, but I doubt that this is still related to the issue.
 - These feeds updated fine in another RSS reader that I was using previously when I enabled the option to ignore self-signed certificates.
 - I saw the issues that seem somewhat related like #3029 but I would like to avoid making these security changes global as a workaround. I just want to disable SSL error checking for specific feeds so that the validity of other feeds is still checked.

**Comment by Frenzie** [comment]
Did you check if this has to do with self-signed as opposed to options along the general lines of `-sha256 -newkey rsa:4096`? (Or at the very least  `-sha256 -newkey rsa:2048`.) The error message doesn't seem to suggest self-signed is the issue, imho.

The above suggestion is based on what I quoted in the issue you linked:

> This is caused by the SECLEVEL 2 setting the security level to 112 bit. This means that RSA and DHE keys need to be at least 2048 bit long. SHA-1 is no longer supported for signatures in certificates and you need at least SHA-256. Note that CAs have stopped issuing certificates that didn't meet those requirements in January 2015, and since January 2017 all valid CA certificates should meet those requirements. However there are certificates generated by private CAs or that are in a test suite that do not meet those requirements.

Since you control the certificates, going there first is probably not a bad idea.

**Comment by Poorchop** [comment]
I don't know of any other feeds with invalid certificates to test against, so I don't know how to test if this is an issue specifically with using a self-signed certificate. According to what you wrote, the algorithm that I used should be good enough even for SECLEVEL 2. I don't really understand the error message generated by FreshRSS but it almost seems like there was an error setting SECLEVEL to 1 for the feeds in question. All installation checks have passed however so it doesn't seem like this could be an issue with write permissions.

Considering that my previous curl issues stemmed from not copying the appropriate files to the chroot, I mentioned that as a possible idea. The other issues (#44 and #502) mentioning self-signed certificates made no mention of any resolution and the error message in #502 was different as it appeared to stem from the user not adding a proper DNS name when generating his certificate.

The same self-signed cert worked fine under miniflux when I enabled the option to use self-signed certs for the feed, but that was also a completely different environment (CentOS 7, PHP not running in a chroot, etc.).

**Comment by Poorchop** [comment]
Looking into this a bit more, the setting on [this line](https://github.com/FreshRSS/FreshRSS/blob/eb5f3a5cf22ae7889dfafb769a4cfead1a82a7e6/lib/lib_rss.php#L188) doesn't make much sense to me. I would imagine that `CURLOPT_SSL_CIPHER_LIST` should be set to a list of ciphers to be used by curl. Setting "Verify SSL security" to "No" for working feeds breaks them, so I don't think that this option is working as intended unless I'm completely misunderstanding.

It would make more sense if this setting made the following changes in [this line](https://github.com/FreshRSS/FreshRSS/blob/eb5f3a5cf22ae7889dfafb769a4cfead1a82a7e6/app/Models/Entry.php#L390) instead:

```php
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
```

This might not be ideal or particularly secure but it seems like it would allow feeds with self-signed certificates to work. A better alternative would be to allow the user to supply a path to the self-signed cert but I think that's more complicated and it would involve more steps on the user's part.

**Comment by Frenzie** [comment]
For better or worse, that's already done on the lines right above?

https://github.com/FreshRSS/FreshRSS/blob/eb5f3a5cf22ae7889dfafb769a4cfead1a82a7e6/app/Models/Entry.php#L387-L391

`curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, $attributes['ssl_verify'] ? true : false); ` is the same thing as

```php
 if (!$attributes['ssl_verify']) { 
 	curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false); 
 } else
 	curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true); 
 }
```

;-)

**Comment by Frenzie** [comment]
But I assume setting the default manually when true is probably a bit redundant, in which case it might be better simplified the way you suggest. Pinging @Alkarex

**Comment by Poorchop** [comment]
My bad, I somehow stared directly at those lines and didn't process them. In any case, I think that my error stems from trying to set `CURLOPT_SSL_CIPHER_LIST` to `DEFAULT@SECLEVEL=1` rather than to a list of ciphers but maybe I'm wrong. I'll have to set up a test environment at some point to see if the feeds work with that line removed. I think that this should be easy to replicate by anyone though—just set Verify SSL security to no for any feed and it will break. That's what happens on my end.

**Comment by Frenzie** [comment]
DEFAULT@SECLEVEL=1 is a cipher list. But perhaps the other end doesn't like it?

**Comment by Poorchop** [comment]
You're right. The config should work the way that it is. I don't have any ideas as to why I'm having this issue and looking up this error hasn't yielded anything helpful so far.

Maybe my version of php-openssl can't use this cipher. That's what the [exit code](https://curl.se/docs/manpage.html) would lead me to think.

**Comment by Poorchop** [comment]
Looking at the [man page](https://man.openbsd.org/SSL_CTX_set_cipher_list.3) for setting ciphers in OpenBSD, it doesn't look like there's a way to set a lower security level. I don't even think that there's an equivalent. Whatever feeds were fixed by this setting might just not be reachable on an OpenBSD host but I think that self-signed certs of a certain minimum cryptographic strength would still work fine as long as lines containing `DEFAULT@SECLEVEL=1` were removed/ignored.

I'm not too familiar with PHP but maybe there's a way to do something like

```php
if (!$attributes['ssl_verify']) && strtolower(PHP_OS) !== 'openbsd') { ... }
```

**Comment by Frenzie** [comment]
I think I see the problem now and a quick test seems to confirm, but NB this is very quick/preliminary. SSL_CTX_set_security_level seems to be new in OpenSSL 1.1.0. So instead, something along the lines of:

```php
if (OPENSSL_VERSION_NUMBER >= WHATEVER_IS_1_1_0) {
  // SECLEVEL
} else {
  // not SECLEVEL
}
```

https://www.php.net/manual/en/openssl.constversion.php

**Comment by Poorchop** [comment]
That's definitely the smarter way of handling this then. I guess there's no way to salvage whatever feeds were fixed by setting SECLEVEL=1 but that's fine by me and this is a good compromise because otherwise, there's no easy way to accept self-signed certificates for specific feeds.

Alternatively, maybe "Accept self-signed certificates" could be a separate option altogether that independently manages `CURLOPT_SSL_VERIFYPEER` and `CURLOPT_SSL_VERIFYHOST`.

**Comment by Frenzie** [comment]
> I guess there's no way to salvage whatever feeds were fixed by setting SECLEVEL=1 

I think they don't have to be salvaged because OpenSSL 1.0.x simply uses that by default, assuming I'm interpreting https://github.com/FreshRSS/FreshRSS/pull/3171 correctly.

**Comment by hreintke** [comment]
Hi,
Since a few weeks I have the same issue as described here on some of my feeds.

I am running FresRSS git master. 

I have : 
`[Mon, 11 Oct 2021 16:30:32 +0200] [warning] --- cURL error 60: SSL certificate problem: certificate has expired [https://www.stm32duino.com/app.php/feed]` when using ssl validation option default, 
and 
`[Mon, 11 Oct 2021 16:31:34 +0200] [warning] --- cURL error 59: failed setting cipher list: DEFAULT@SECLEVEL=1 [https://www.stm32duino.com/app.php/feed]` when using ssl validation option no.

Running the php script from https://www.php.net/manual/en/openssl.constversion.php I get 
`openssl version text: OpenSSL 1.1.0l 10 Sep 2019   openssl version number: 269484239`

I don't understand all details from the above discussion. 
Is there a solution from within freshrss or configuring elsewhere ?

**Comment by Frenzie** [comment]
I doubt that's related to the above discussion. See https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/

You will however need to update your packages.

**Comment by hreintke** [comment]
Thx for pointing in the right direction.
I moved my FreshRSS to a system that correctly implements the right certificates. 

So it works for me know but what I still see is that when setting the option ssl validation to "no" there is always the "failed setting cipher list error" 

This is regardless whether I use a feed which had the "certificate issue" or not. 
Setting the ssl validation to no always gives the error.

**Comment by Frenzie** [comment]
I don't think that should happen on OpenSSL 1.1.x. What kind of system are you trying it on?

**Comment by hreintke** [comment]
I have it running now on a synology NAS, with the issue of ssl validation. 
This was updated recently which included a fix for the certificate, although still running on openssl 1.0.2u-fips.

The other is a mediaserver, based on Debian. This is using openssl 1.1.01. But does show the original issue of expired certificate.

I do have a raspberry available, will do some testing on that too.

**Comment by math-GH** [comment]
Is anything left here that need support?

**Comment by Poorchop** [comment]
As far as I'm aware, this was never addressed. I don't remember the exact changes that I made but whenever I update FreshRSS, I believe that I have to comment out some of the aforementioned lines in order for everything to work.

**Comment by Alkarex** [comment]
Related draft for adding more cURL options, help welcome https://github.com/FreshRSS/FreshRSS/pull/4573
